### PR TITLE
TestingWithMocksTaskOne

### PR DESCRIPTION
**Pull request by aleksandrhristov**
EndPointFilter Task !

### CODE REVIEW

**Comment by freekman** [comment]
For the PR you should add Proper Title and Description

File: src/com/clouway/endpointfilter/Endpoint.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.clouway.endpointfilter;
+
+/**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+ * Created by clouway on 07.07.16.
```
Comment (freekman) [inline src/com/clouway/endpointfilter/Endpoint.java]: remove this line

File: src/com/clouway/endpointfilter/EndpointFilter.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.clouway.endpointfilter;
+
+/**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+ */
+public class EndpointFilter {
```
Comment (freekman) [inline src/com/clouway/endpointfilter/EndpointFilter.java]: 1)format class
2) why do you have comments on the constructor and method

File: src/com/clouway/endpointfilter/Endpoint.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.clouway.endpointfilter;
+
+/**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
```
Comment (freekman) [inline src/com/clouway/endpointfilter/Endpoint.java]: the template should be 
```js
* @author Aleksandar Hristov <alexanderhristov93@gmail.com> 
```
without the Email prefix

File: src/com/clouway/endpointfilter/MissingKeywordException.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.clouway.endpointfilter;
+
+/**
+<<<<<<< HEAD
```
Comment (freekman) [inline src/com/clouway/endpointfilter/MissingKeywordException.java]: ?

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package com.clouway.endpointfilter; /**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+ */
+
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: format

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ */
+public class EndpointFilterTest {
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery();
+
+    private Endpoint endpoint = context.mock(Endpoint.class);
+    private EndpointFilter endpointFilter = new EndpointFilter(endpoint);
+
+    @Test
+    public void endPointIsMatching() throws MissingURLException, MissingKeywordException {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: rename to happyPath

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            will(returnValue(true));
+        }});
+        boolean result = endpointFilter.shouldFilter("test");
+        assertTrue(result);
+    }
+
+    @Test
+    public void endPointIsNotMatching() throws MissingURLException, MissingKeywordException {
+        context.checking(new Expectations() {{
+            oneOf(endpoint).matches("ThisUrlDoesNotMatch");
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: use with(any(String.class) like in the upper test

File: src/com/clouway/endpointfilter/Endpoint.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+ * Created by clouway on 07.07.16.
+ */
+// Interface which will be mocked !
```
Comment (freekman) [inline src/com/clouway/endpointfilter/Endpoint.java]: please take a look of how to write documentation in java

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            will(returnValue(false));
+        }});
+        EndpointFilter endpointFilter = new EndpointFilter(endpoint);
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: why are you re-instantiating EndpointFilter

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        assertFalse(endpointFilter.shouldFilter("ThisUrlDoesNotMatch"));
+    }
+
+    @Test
+    public void endPointMany() throws MissingKeywordException, MissingURLException {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: test ```endPointMany``` what is the purpose of this test.

File: src/com/clouway/endpointfilter/MissingKeywordException.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+=======
+ * Created by clouway on 07.07.16.
+>>>>>>> 3ccf7a7d953d1497ab9a8cfe392128784a167197
+ */
+public class MissingKeywordException extends RuntimeException{
+    public MissingKeywordException() {
```
Comment (freekman) [inline src/com/clouway/endpointfilter/MissingKeywordException.java]: ?

File: src/com/clouway/endpointfilter/MissingURLException.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package com.clouway.endpointfilter;
+
+/**
+<<<<<<< HEAD
```
Comment (freekman) [inline src/com/clouway/endpointfilter/MissingURLException.java]: .

File: src/com/clouway/endpointfilter/MissingURLException.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+=======
+ * Created by clouway on 07.07.16.
+>>>>>>> 3ccf7a7d953d1497ab9a8cfe392128784a167197
+ */
+public class MissingURLException extends RuntimeException {
+    public MissingURLException() {
```
Comment (freekman) [inline src/com/clouway/endpointfilter/MissingURLException.java]: .

File: src/com/clouway/endpointfilter/StartsWithKeyword.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
+package com.clouway.endpointfilter;
+
+/**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+ */
+
+class StartsWithKeyword implements Endpoint {   // implements the interface Endpoint
```
Comment (freekman) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: why do you have a comment on every line

File: src/com/clouway/endpointfilter/StartsWithKeyword.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
+package com.clouway.endpointfilter;
+
+/**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
+ */
+
```
Comment (freekman) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: I wont write comments for every empty line please format your code

**Comment by aleksandrhristov** [comment]
all coments fixed

**Comment by aleksandrhristov** [comment]
and for sure  i will write documentation on saturday and on sunday when you cant check my tasks

File: src/com/clouway/endpointfilter/StartsWithKeyword.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+class StartsWithKeyword implements Endpoint {   // implements the interface Endpoint
+    private final String keyWord; //field for keyword
+    //adding to the constructor Keyworrd
+    public StartsWithKeyword(String keyWord) {
+        this.keyWord = keyWord;
+    }
+    //overriding matches      if url is empty throw exec if keyword is empty throw exec if url stars with keyword return true else false ;
+//overriding matches      if url is empty throw exec if keyword is empty throw exec if url stars with keyword return true else false ;
+    @Override
+    public boolean matches(String url) {
```
Comment (freekman) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: why does a method ```boolean matches() ``` throw an exception

I would want to read and tell me where it is appropriate to throw exceptions

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Created by clouway on 07.07.16.
+ */
+public class StartsWithKeywordTest {
+    @Test
+    public void endPointIsMatching() throws MissingURLException, MissingKeywordException {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: matches

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Test(expected = MissingURLException.class)
+    public void emptyUrl() throws MissingURLException, MissingKeywordException {
+        StartsWithKeyword startsWithKeyword = new StartsWithKeyword("Start");
+        startsWithKeyword.matches("");
+    }
+
+    @Test
+    public void endPointDoesNotMatch() throws MissingURLException, MissingKeywordException {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: notmatches

File: src/com/clouway/endpointfilter/StartsWithKeyword.java (hunk 16 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: i delete this

**Comment by aleksandrhristov** [comment]
fixed comments

**Comment by freekman** [comment]
Not all the comments are fixed

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
+package com.clouway.endpointfilter;
+
+/**
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: .

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author Aleksandar Hristov <Email:alexanderhristov93@gmail.com>
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: .

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package com.clouway.endpointfilter;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Created by clouway on 07.07.16.
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: .

File: src/com/clouway/endpointfilter/Endpoint.java (hunk 1 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/Endpoint.java]: done

File: src/com/clouway/endpointfilter/EndpointFilter.java (hunk 2 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/EndpointFilter.java]: removed comments done and format class

File: src/com/clouway/endpointfilter/Endpoint.java (hunk 3 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/Endpoint.java]: done

File: src/com/clouway/endpointfilter/MissingKeywordException.java (hunk 4 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingKeywordException.java]: done

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 5 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: done

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 6 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: done

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 7 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: in endpointfiltertest i replace them all

File: src/com/clouway/endpointfilter/Endpoint.java (hunk 8 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/Endpoint.java]: can i write documentation Saturday and Sunday?

File: src/com/clouway/endpointfilter/MissingKeywordException.java (hunk 11 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingKeywordException.java]: done

File: src/com/clouway/endpointfilter/MissingURLException.java (hunk 12 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingURLException.java]: done

File: src/com/clouway/endpointfilter/MissingURLException.java (hunk 13 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingURLException.java]: done

File: src/com/clouway/endpointfilter/MissingKeywordException.java (hunk 11 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingKeywordException.java]: done

File: src/com/clouway/endpointfilter/MissingURLException.java (hunk 12 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingURLException.java]: done

File: src/com/clouway/endpointfilter/MissingURLException.java (hunk 13 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/MissingURLException.java]: done

File: src/com/clouway/endpointfilter/StartsWithKeyword.java (hunk 16 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: done

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java (hunk 17 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: done

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java (hunk 21 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: done

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java (hunk 20 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: done

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 19 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: done

Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: the author of a class must be above the ```+public class EndpointFilterTest {```

File: src/com/clouway/endpointfilter/EndpointFilter.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+    private final Endpoint[] endpoints;
+
+    public EndpointFilter(Endpoint... endpoints) {
+        this.endpoints = endpoints;
+    }
+
+    public boolean shouldFilter(String url) {
+        for (Endpoint each : endpoints) {
```
Comment (freekman) [inline src/com/clouway/endpointfilter/EndpointFilter.java]: rename each to endpoint

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+        context.checking(new Expectations() {{
+            oneOf(endpoint).matches(with(any(String.class)));
+            will(returnValue(false));
+        }});
+        boolean result = endpointFilter.shouldFilter("test");
+        assertFalse(result);
+    }
+
+    @Test
+    public void endPointMany() {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: multipleFitrations

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+        final Sequence matchEndpoints = context.sequence("MatchEndpoints");
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: why are you using this

File: test/com/clouway/endpointfilter/EndpointFilterTest.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+        context.checking(new Expectations() {{
+            oneOf(endpoint).matches(with(any(String.class)));
+            will(returnValue(true));
+        }});
+        boolean result = endpointFilter.shouldFilter("test");
+        assertTrue(result);
+    }
+
+    @Test
+    public void endPointNotMatches() {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: are you testing the endpoint not matches or the should filter

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+ */
+public class StartsWithKeywordTest {
+    @Test
+    public void endPointIsMatching() {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: matches

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+        StartsWithKeyword startsWithKeyword = new StartsWithKeyword("Start");
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: extract this ``` StartsWithKeyword startsWithKeyword = new StartsWithKeyword("Start");``` to reuse it in all the tests

File: src/com/clouway/endpointfilter/EndpointFilter.java (hunk 22 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/EndpointFilter.java]: done

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+        assertTrue(startsWithKeyword.matches("Start with"));
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: extract this in a variable ```startsWithKeyword.matches("Start with")``` for beter readablity

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java (hunk 26 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: done

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 25 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: the filter

File: test/com/clouway/endpointfilter/StartsWithKeywordTest.java [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Test
+    public void endPointNotMatch() {
```
Comment (freekman) [inline test/com/clouway/endpointfilter/StartsWithKeywordTest.java]: think of a betar name

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 25 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: i rename

File: src/com/clouway/endpointfilter/StartsWithKeyword.java [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+class StartsWithKeyword implements Endpoint {
+    private final String keyWord;
+
+    public StartsWithKeyword(String keyWord) {
+        this.keyWord = keyWord;
+    }
+
+    @Override
+    public boolean matches(String url) {
+        if (url.isEmpty()) {
```
Comment (freekman) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: missing test casses for 
```js
        if (url.isEmpty()) {
 +            return false;
 +        }
 +        if (keyWord.isEmpty()) {
 +            return false;
 +        }
```

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 24 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: done i think

File: test/com/clouway/endpointfilter/EndpointFilterTest.java (hunk 23 above)
Comment (aleksandrhristov) [inline test/com/clouway/endpointfilter/EndpointFilterTest.java]: done

File: src/com/clouway/endpointfilter/StartsWithKeyword.java (hunk 30 above)
Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/StartsWithKeyword.java]: i write exceptions and they was wrong or not ?

File: src/com/clouway/endpointfilter/Endpoint.java [hunk 31]
```diff
@@ -0,0 +1,15 @@
+package com.clouway.endpointfilter;
+
+/**
+ * @author Aleksandar Hristov <alexanderhristov93@gmail.com>
+ */
+
+/**
+ * Comparison the url.
+ *@param url the first number to sum
+ *@return url is matches or not matches (true or false)(boolean)
```
Comment (freekman) [inline src/com/clouway/endpointfilter/Endpoint.java]: are you kidding me

Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/Endpoint.java]: my bad

Comment (aleksandrhristov) [inline src/com/clouway/endpointfilter/Endpoint.java]: fixing

**Pull request closed without merging**